### PR TITLE
test(NODE-4296): add csfle decryption events prose tests

### PR DESCRIPTION
**Pull request by durran**
### Description

Add the decryption events prose tests for CSFLE.

#### What is changing?

4 new prose tests have been added to test command monitoring when using CSFLE in various cases. I've copied and pasted the prose tests as comments in the code. The new prose tests language can also be referenced here: https://github.com/mongodb/specifications/commit/f9cbedafe53faffd62f15bcb2e952d1e016934e9

##### Is there new documentation needed for these changes?

None

#### What is the motivation for this change?

NODE-4296

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js [hunk 1]
```diff
@@ -1424,4 +1424,245 @@ describe('Client Side Encryption Prose Tests', metadata, function () {
...
+        const collection = encryptedClient.db('').collection('decryption_events');
+        let result;
+        try {
+          result = await collection.aggregate([]).toArray();
+        } catch (error) {
+          expect(error.code).to.equal(123);
+        }
+        expect(result).to.not.exist;
+        expect(aggregateFailed.failure.code).to.equal(123);
+        expect(aggregateStarted).to.exist;
```
Comment (durran) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:1567]: The prose tests say to keep the command started events for the aggregations but never asks to assert anything about them, so I just assert they are emitted.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js [hunk 2]
```diff
@@ -1424,4 +1424,245 @@ describe('Client Side Encryption Prose Tests', metadata, function () {
...
+        const collection = encryptedClient.db('').collection('decryption_events');
+        await collection.drop();
+        await collection.insertOne({ encrypted: cipherText });
+        let result;
+        try {
+          result = await collection.aggregate([]).toArray();
+        } catch (error) {
+          expect(error).to.not.exist;
+        }
+        expect(result[0].encrypted).to.equal('hello');
```
Comment (durran) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:1661]: I actually assert the value of the field here is the original value decrypted, and then in line 1663 is the important part showing the same field in the command succeeded event is still an encrypted `Binary`

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js [hunk 3]
```diff
@@ -1424,4 +1424,245 @@ describe('Client Side Encryption Prose Tests', metadata, function () {
...
+        // to contain BSON binary for the field
+        // ``cursor.firstBatch.encrypted``.
+        const collection = encryptedClient.db('').collection('decryption_events');
+        await collection.drop();
+        await collection.insertOne({ encrypted: malformedCiphertext });
+        let result;
+        try {
+          result = await collection.aggregate([]).toArray();
+        } catch (error) {
+          expect(error.message).to.include('HMAC validation failure');
```
Comment (durran) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:1635]: This is basically showing that we generate command monitoring events before we decrypt as we error during decryption but still get a command succeeded event. Kind of funny but the most secure way to do it I guess.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js [hunk 4]
```diff
@@ -1424,4 +1424,243 @@ describe('Client Side Encryption Prose Tests', metadata, function () {
...
+      // Create a MongoClient named ``encryptedClient`` with these ``AutoEncryptionOpts``:
+      //   AutoEncryptionOpts {
+      //     keyVaultNamespace: "keyvault.datakeys";
+      //     kmsProviders: { "local": { "key": <base64 decoding of LOCAL_MASTERKEY> } }
+      //   }
+      // Configure ``encryptedClient`` with "retryReads=false".
+      encryptedClient = this.configuration.newClient(
+        {},
+        {
+          retryReads: false,
```
Comment (baileympearson) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:1489]: question: is there a reason that retryReads is set to false here?  I don't see that listed as a part of the prose test setup.  Is this just related to FLE in general?

Comment (baileympearson) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js:1489]: nevermind, it was the next line in the test setup 😅

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js [hunk 5]
```diff
@@ -1424,4 +1424,243 @@ describe('Client Side Encryption Prose Tests', metadata, function () {
...
+        // Use ``encryptedClient`` to run an aggregate on ``db.decryption_events``.
+        // Expect an exception to be thrown from the command error. Expect a CommandFailedEvent.
+        const collection = encryptedClient.db('').collection('decryption_events');
+        let result;
+        try {
+          result = await collection.aggregate([]).toArray();
+        } catch (error) {
+          expect(error.code).to.equal(123);
+        }
+        expect(result).to.not.exist;
```
Comment (baileympearson) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:1565]: this is a nitpick so feel free to ignore it - rather than asserting that result is undefined, we could use an `expect.fail` between lines 1561 and 1562.  That will fail the test if an error isn't thrown by `collection.aggregate`.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js (hunk 1 above)
Comment (baileympearson) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js:1567]: It is strange that the prose tests say nothing about this 🤷‍♂️

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js [hunk 6]
```diff
@@ -1424,4 +1424,243 @@ describe('Client Side Encryption Prose Tests', metadata, function () {
...
+
+      it('expects an error and a command failed event', async function () {
+        // Use ``encryptedClient`` to run an aggregate on ``db.decryption_events``.
+        // Expect an exception to be thrown from the network error. Expect a CommandFailedEvent.
+        const collection = encryptedClient.db('').collection('decryption_events');
+        let result;
+        try {
+          result = await collection.aggregate([]).toArray();
+        } catch (error) {
+          expect(error.message).to.include('closed');
```
Comment (baileympearson) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:1611]: could we assert that the error is an instanceof `MongoNetworkError`?

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js [hunk 7]
```diff
@@ -1424,4 +1424,243 @@ describe('Client Side Encryption Prose Tests', metadata, function () {
...
+        // Expect no exception.
+        // Expect a CommandSucceededEvent. Expect the CommandSucceededEvent.reply
+        // to contain BSON binary for the field ``cursor.firstBatch.encrypted``.
+        const collection = encryptedClient.db('').collection('decryption_events');
+        await collection.insertOne({ encrypted: cipherText });
+        let result;
+        try {
+          result = await collection.aggregate([]).toArray();
+        } catch (error) {
+          expect(error).to.not.exist;
```
Comment (baileympearson) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:1657]: again, just a minor nit but we could use expect.fail here too.  

I kinda see `expect.fail` as being used to guard against branches in logic that should never be hit, and the regular assertion statements to assert on values in the expected path.

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js (hunk 6 above)
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Changed to assert error type.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js (hunk 5 above)
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Updated to use `expect.fail`

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js (hunk 1 above)
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js:1563]: I just realised I was looking at the commit for the changes which mentions that, but it got updated to remove it in the most recent. I'll remove these now.

**Review by baileympearson — APPROVED** [review]

**Review by dariakp — COMMENTED** [review]
Just have a question here, otherwise lgtm

**Review by dariakp — APPROVED** [review]
